### PR TITLE
Streaming search: stream filters

### PR DESCRIPTION
**Pull request by camdencheek**
Currently, we don't send filters until a search has completed. This means that we don't render the sidebar dynamic filters until the end of the search. This leads to a somewhat jarring page flashing where 1) you start a search, 2) the first results show up, 3) the search completes and the dynamic filters load. 

This updates the backend to stream filters instead. Currently, it computes and sends filters whenever we flush matches. I don't think this will be practically too expensive, given the relatively low cardinality of repos compared to matches, but I'll keep an eye on profiles in prod, but it seems fine locally. If it is too expensive, I'm pretty sure we can compute filters incrementally as we stream. 

Additionally, this fixes an rxjs issue I noticed with `throttleTime`. Basically, with leading and trailing set for `throttleTime`, it no longer guarantees a minimum separation between output events. This causes a "galloping" effect in the UI where you'll see pairs of very close updates. It looks very unnatural. 

## Test plan

Manually tested. See attached video for results. 

https://user-images.githubusercontent.com/12631702/156481507-eeffad75-af53-4136-977b-5b3e990d50d4.mov

### CODE REVIEW

**Comment by camdencheek** [comment]
For the frontend people: in the video, it's very noticeable that the scrollbar pops up when the number of dynamic filters changes. Is there anything we can do about that?

**Comment by camdencheek** [comment]
The other thing: is this something that people actually like? I'm not sure the extra movement on the page is a positive thing, but I also _do_ like that the dynamic filters render immediately on the first result.

**Comment by limitedmage** [comment]
> For the frontend people: in the video, it's very noticeable that the scrollbar pops up when the number of dynamic filters changes. Is there anything we can do about that?

Do you mean that when new filters are added, the scroll position resets? There are ways to preserve the scroll position. We can look into optimizing this on the UI side. 

If you mean that the scrollbar shows up when new items are added, this is greatly dependent on which browser/OS you're using and it's something we can't really control without a lot of usability/accessibility drawbacks. TBH I don't think the "popping up" that shows in the video is all that bad. What browser/OS are you using? I can't see the popping locally macOS 12 with any of the different scrollbar settings in System Preferences > General), on either Chrome or Safari, but I do see it in Firefox (only when scrollbars set to "when scrolling"). So it might be a bug in Mac Firefox only.

**Comment by camdencheek** [comment]
> but I do see it in Firefox

Yep, this is Firefox on Mac

File: client/web/src/search/results/SearchResultsCacheProvider.tsx [hunk 1]
```diff
@@ -55,9 +55,15 @@ export function useCachedSearchResults(
...
-            return streamSearch(transformedQuery, options).pipe(
-                throttleTime(500, undefined, { leading: true, trailing: true })
-            )
+            const stream = streamSearch(transformedQuery, options)
+
+            // If the throttleTime option `trailing` is set, we will return the
+            // final value, but it also removes the guarantee that the output events
+            // are a minimum of 200ms apart. If it's unset, we might throw away
+            // some trailing events. This is a fundamental issue with throttleTime,
+            // and is discussed extensively in github issues. Instead, we just manually
```
Comment (limitedmage) [inline client/web/src/search/results/SearchResultsCacheProvider.tsx:64]: Weird and good find! Can you add a link to one of the GitHub issues as a comment, ideally in the rxjs repo itself? (For future reference to whoever reads this comment years from now and wants to check if the bug has been resolved)

**Review by limitedmage — APPROVED** [review]

**Pull request merged**